wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/environment/server/CheEnvironmentEngine.java

```diff
@@ -861,29 +1017,62 @@ private void ensurePreDestroyIsNotExecuted() throws ServerException {
         }
     }
 
+    private ComposeServiceImpl machineConfigToComposeService(MachineConfig machineConfig) throws ServerException {
+        ComposeServiceImpl composeService = new ComposeServiceImpl();
+        composeService.setMemLimit(machineConfig.getLimits().getRam() * 1024L * 1024L);
+        composeService.setEnvironment(machineConfig.getEnvVariables());
+        if ("image".equals(machineConfig.getSource().getType())) {
+            composeService.setImage(machineConfig.getSource().getLocation());
+        } else {
+            if (machineConfig.getSource().getContent() != null) {
+                throw new ServerException(
+                        "Additional machine creation from dockerfile content is not supported anymore. " +
+                        "Please use dockerfile location instead");
+            } else {
+                composeService.setBuild(new BuildContextImpl(machineConfig.getSource().getLocation(), null));
+            }
+        }
+        List<? extends ServerConf> servers = machineConfig.getServers();
+        if (servers != null) {
```

Comment:
afaik it can't be null

```java
composeService.setExpose(machineConfig.getServers();
                                      .map(ServerConf::getPort)
                                      .collect(toList());
```